biz.aQute.tester.junit-platform/src/aQute/tester/junit/platform/Activator.java

```diff
@@ -0,0 +1,377 @@
+		} else {
+			trace("using %s, path: %s", reportDir, reportDir.toPath());
+			try {
+				listenerList
+					.add(new LegacyXmlReportGeneratingListener(reportDir.toPath(), new PrintWriter(System.err)));
+			} catch (Exception e) {
+				error("Error trying to create xml reporter: %s", e);
+			}
+		}
+
+		basicReport = LoggingListener.forBiConsumer((t, msg) -> {
+			if (t == null) {
+				trace(msg.get());
+			} else {
+				trace(msg.get(), t);
+			}
+		});
+		summary = new SummaryGeneratingListener();
+		listenerList.add(basicReport);
+		listenerList.add(summary);
+		listeners = listenerList.stream()
+			.toArray(TestExecutionListener[]::new);
+
+		if (testcases == null) {
+			trace("automatic testing of all bundles with " + aQute.bnd.osgi.Constants.TESTCASES + " header");
+			try {
+				automatic();
+			} catch (IOException e) {
+				// ignore
+			}
+		} else {
+			trace("receivednames of classes to test %s", testcases);
+			try {
+				baseSelectors = Stream.of(testcases.split("\\s*,\\s*"))
+					.map(x -> {
+						if (x.contains(":")) {
+							return selectMethod(x.replace(':', '#'));
+						} else {
+							return selectClass(x);
+						}
+					})
+					.collect(Collectors.toList());
+				automatic();
+			} catch (Exception e) {
+				e.printStackTrace();
+				System.exit(254);
+			}
+		}
+	}
+
+	void automatic() throws IOException {
+		final BlockingDeque<LauncherDiscoveryRequest> queue = new LinkedBlockingDeque<>();
+
+		trace("adding Bundle Listener for getting test bundle events");
+		context.addBundleListener((SynchronousBundleListener) event -> {
+			switch (event.getType()) {
+				case BundleEvent.STARTED :
+				case BundleEvent.RESOLVED :
+					checkBundle(queue, event.getBundle());
+					break;
+			}
+		});
+
+		long result = 0;
+		LauncherDiscoveryRequest request = buildRequest();
+		TestPlan plan = launcher.discover(request);
+		if (plan.containsTests()) {
+			result = test(request);
+			if (!continuous) {
+				System.exit((int) result);
+			}
+		}
+
+		trace("starting queue");
+		while (active()) {
+			try {
+				LauncherDiscoveryRequest testRequest = queue.takeFirst();
+				trace("test will run");
+				result += test(testRequest);
+				trace("test ran");
+				if (queue.isEmpty() && !continuous) {
+					trace("queue " + queue);
+					System.exit((int) result);
+				}
+			} catch (InterruptedException e) {
+				trace("tests bundle queue interrupted");
+				thread.interrupt();
+			} catch (Exception e) {
+				error("Not sure what happened anymore %s", e);
+				System.exit(254);
+			}
+		}
+	}
+
+	List<DiscoverySelector> baseSelectors = Collections.emptyList();
+
+	LauncherDiscoveryRequest buildRequest() {
+		return LauncherDiscoveryRequestBuilder.request()
+			.configurationParameter(BundleEngine.CHECK_UNRESOLVED, unresolved)
+			.selectors(baseSelectors)
+			.build();
+	}
+
+	LauncherDiscoveryRequest buildRequest(Bundle bundle) {
+		List<DiscoverySelector> selectors = new ArrayList<>(baseSelectors.size() + 1);
```

Comment:
Shouldn't this copy in `baseSelectors`? I know it it currently empty, but the code implies you want to add another selector to the base selectors.

Reply:
Well spotted. This is a good question.

I think that my original intent was to copy `baseSelectors` in. I think in practice when I started using it it was annoying me, as I was trying to re-run the tests for the whole bundle and it kept re-running only the classes/methods that was originally included in `-tester.names` (or indeed not running anything if the bundle in question didn't contain the classes/methods in question). I think in combination with #3320 it was causing me some grief. I probably removed the copy during my testing and forgot to put it back in.

I'm open for suggestions here as to what the best behaviour is. I'm thinking that once #3320 is resolved, then copying the base selectors probably makes sense.

I also thought at some point to extend the `-tester.names` syntax to support `bundle`, `bundle:class` and `bundle:class:method` syntax, but to be honest I just wanted to get it merged so that this could be added as an additional feature later.

Reply:
So either get rid of baseSelectors for now, or copy it into the new array. The current state is too much in the middle.

Reply:
I've copied on baseSelectors for now.